### PR TITLE
New Wesh blog "Flight of a byte"

### PR DESCRIPTION
**Pull request by iuricmp**
Signed-off-by: Iuri Pereira <iuricmp@gmail.com>

### CODE REVIEW

File: pages/posts/flight-of-a-byte.tsx [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+                </p>
+                <p>
+                When the Wesh app on the computer for <code>client2</code> sends the bytes of a message, what is the <code className="language-go">“flight path”</code> of the bytes of <code className="language-go">“Hello”</code> as they travel to the computer of <code>client1</code>? It&apos;s different from “traditional” networking where, for example, a computer sends the message in a UDP packet to the IP address of the other computer, as described in this Wikipedia article. Instead of a specific IP address, Wesh communication happens between libp2p peers, where the client&apos;s PeerID is the value that we&apos;ve seen in previous examples and is the same no matter how the client is connected to the network.
```
Comment (LFGaming) [inline pages/posts/flight-of-a-byte.tsx:38]: No wikipedia article is linked here

File: pages/posts/content.json [hunk 2]
```diff
@@ -1,5 +1,15 @@
 {
   "items": [
+    {
+      "id": "5",
+      "title": "5 - Wesh: Flight of a Byte",
+      "pubDate": "20/07/2023",
```
Comment (jefft0) [inline pages/posts/content.json:6]: At the bottom of previous blog posts, it says something like "Posted by Jeff on 20/07/2023". Should that be added automatically by this metadata? In any case, the current blog post does not have a "posted" date. Can you add one?

File: pages/posts/flight-of-a-byte.tsx (hunk 1 above)
Comment (jefft0) [reply pages/posts/flight-of-a-byte.tsx:38]: @iuricmp , please add the link from the Notion draft.

Comment (jefft0) [reply pages/posts/flight-of-a-byte.tsx:38]: These should be a link for "this Wikipedia article", and another link for "libp2p peers".

File: pages/posts/flight-of-a-byte.tsx [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+                </p>
+                <h3>Peer discovery</h3>
+                <p>
+                The “glue” between a libp2p peers and a more traditional network address is <a href="https://docs.libp2p.io/concepts/discovery-routing/overview/">peer discovery</a>. There are a lot of details that we won&apos;t go into regarding types of discovery, peer authentication, etc. In this blog post, we just describe the minimal steps for how <code>client1</code> receives the message from <code>client2</code>. When the libp2p service of client1 discovers the computer of <code>client2</code>, it opens a network connection (<a href="https://www.notion.so/Backlogs-for-1-Berty-Messenger-Launch-and-2-Wesh-Network-d420a66eea244942878b31d3fc6eab4a?pvs=21">code</a>) and reads from it using a traditional byte stream (<a href="https://github.com/libp2p/go-libp2p-pubsub/blob/6d73cd4b56bf9d7fdd0c0caaed629a3d51027a75/comm.go#L66">code</a>). (We give links to code for curious readers, but this isn&apos;t a code tutorial and you can skip the code links.)
+
```
Comment (jefft0) [inline pages/posts/flight-of-a-byte.tsx:43]: In "network connection (code)", I put the wrong link in the Notion page. Sorry. (It shouldn't be a link to our notion site.) Please change the link to:
https://github.com/libp2p/go-libp2p-pubsub/blob/6d73cd4b56bf9d7fdd0c0caaed629a3d51027a75/comm.go#L115

File: pages/posts/flight-of-a-byte.tsx [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+                We&apos;ve mentioned the traditional networking code “under the hood” to remove some of the mystery. But that code doesn&apos;t immediately read the message from <code>client2</code>. Instead, there are some communication steps which allow Wesh to use libp2p to operate in a peer-to-peer environment.
+                </p>
+                
+                <h3>Message log and “head”</h3>
+                <p>
+                To send a message in traditional networking, one user connects to the network address of the other user and transmits the message. Or the user connects to a central server and appends to the message log there, while the other user connects to the same central server and reads the message log. But this doesn&apos;t work in a peer-to-peer environment. Instead, the goal is for each group member to have a copy of the message log, where there is a “head” (the latest message) which refers to all the previous messages in a chain. To send a message to the group, a member adds to the local copy of the message log and updates the “head”. This will be synchronized with the other members.
+                </p> 
+                <p>
```
Comment (jefft0) [inline pages/posts/flight-of-a-byte.tsx:51]: This is missing a paragraph. Please include the paragraph from the Notion page starting with "Without a central server..."

File: pages/posts/flight-of-a-byte.tsx [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+                 As mentioned, <code>client2</code> doesn&apos;t send the message “Hello” directly to the other group members. Instead, it puts the message in its own message log and updates the head. The other members learn about the new head which means there is a new message to fetch.
+                </p>
+                <h3>libp2p pubsub</h3>
+                <p>
+                 So now the question is, how does <code>client1</code> learn that <code>client2</code> says there is a new message log head? This uses libp2p pubsub where peers can subscribe to a topic and publish a pubsub message to that topic, or receive pubsub messages. (We say “pubsub message” to not confuse with the ordinary messages in the log which we&apos;ll discuss below.)
+                </p>
+                <p>
+                libp2p pubsub is a powerful tool with algorithms which allow it to scale to many peers and to make sure they all get the pubsub message. You can read more in its documentation, but for this blog post we only need to say that the Wesh group has a libp2p topic name like 
+                &nbsp;<code>
```
Comment (jefft0) [inline pages/posts/flight-of-a-byte.tsx:60]: The Notion page has a link for "its documentation". Please include the link.

File: pages/posts/flight-of-a-byte.tsx [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+                /orbitdb/bafyreihpdptqwgpuwu4ob6nusxebmdx6fm6mq5pma5hva6eu5u74pflo7e/2ac4e88c5272f900c76fd36989f9780e5b2c95d75d38fe5bcef0345b34bc4806_wesh_group_messages
+                    </code>
+                &nbsp;a pubsub message to this topic that contains the message log head. The libp2p on <code>client1</code> receives this pubsub message and sends it (code) internally to all processes which are subscribed to the topic. The Wesh application of <code>client1</code> has used ActivateGroup and is therefore subscribed to this topic and receives the pubsub message (code).
+                </p>
```
Comment (jefft0) [inline pages/posts/flight-of-a-byte.tsx:64]: "a pubsub message to this topic" should be "Client2 publishes a pubsub message to this topic."

Comment (jefft0) [reply pages/posts/flight-of-a-byte.tsx:64]: Also, please include the code links for "sends it (code)" and "pubsub message (code)". See the Notion draft.

File: pages/posts/flight-of-a-byte.tsx [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+                <h3>IPFS to fetch the message</h3>
+                <p>
+                libp2p pubsub doesn&apos;t care about the meaning of a pubsub message. It only makes sure that all subscribers get it. Remember that the next step is to synchronize the message log, which is handled by the local message store of <code>client1</code>. After checking that it&apos;s not a repeat, the pubsub message is forwarded (code) as an EventPubSubMessage to the message store which receives it (code). This confirms that the type of pubsub message is to update the message log head, and it checks to make sure the head was not already processed. (If <code>client1</code> has already received the same message log head, maybe from another peer, then it doesn&apos;t need to do more processing. This may seem trivial, but it&apos;s important because it&apos;s efficient to exchange these small “head” messages frequently between peers.)  
+                </p>
```
Comment (jefft0) [inline pages/posts/flight-of-a-byte.tsx:68]: Please include the code links for "is forwarded (code)" and "receives it (code)".

File: pages/posts/flight-of-a-byte.tsx [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+                <p>
+                The message store receives (code) the event for a new head which contains the <code>CID</code> of the message. The message (which may be large) is fetched with IPFS, not with libp2p pubsub (which is better for small messages). If you don&apos;t know what a <code>CID</code> is and how IPFS uses it to fetch content, then you can read all about it. Notice that using a <code>CID</code> to fetch the message (code) like “Hello” is a separate process from synchronizing heads. Maybe IPFS has already fetched the same <code>CID</code>, or maybe it can be fetched from another group member that is not <code>client2</code>. Wesh takes advantage of the efficient IPFS algorithms.
+                </p>
```
Comment (jefft0) [inline pages/posts/flight-of-a-byte.tsx:71]: Please include the code links for "receives (code)" and "fetch the message (code)". Also please include the link for "read all about it".

File: pages/posts/flight-of-a-byte.tsx [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+                <h3>Decrypting and finishing</h3>
+                <p>
+                We&apos;re almost done! When the “Hello” message is fetched, it&apos;s placed in the local copy of the message log in the correct order. (Messages in the device&apos;s storage remain encrypted.) This internally sends the EventReplicated event which is received (code) by the Wesh message monitor. We started at the libp2p “level” which doesn&apos;t care about the meaning of pubsub messages, up through the message log level which only cares about the correct order of messages. Now at the “top” level, Wesh must use <code>client1&apos;s</code> cryptographic keys to decrypt (code) the message.
+                </p>
```
Comment (jefft0) [inline pages/posts/flight-of-a-byte.tsx:75]: Please include code links.

File: pages/posts/flight-of-a-byte.tsx [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+                <p>
+                The decrypted message is sent internally (code) to all monitoring processes, including the GroupMessageList service (code). The <code>client1</code> application has called GroupMessageList, so the “Hello” message is sent (code) to the application which prints the message.
+                </p>
```
Comment (jefft0) [inline pages/posts/flight-of-a-byte.tsx:78]: Please include the code links.

File: pages/posts/flight-of-a-byte.tsx [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+                <h3>Flight of a Byte</h3>
+                <p>
+                In summary, the flight of a message from <code>client2</code> to client1 starts when the <code>client2</code> peer is discovered, subscribes to the libp2p pubsub topic for the group messages and publishes its message log head. <code>client1</code> doesn&apos;t have this head yet, so it uses the CID to fetch the message using IPFS. When the message arrives, it is added to <code>client1&apos;s</code> copy of the message log in the correct order. Wesh sends this new message to processes using the GroupMessageList service, such as the sample application which receives it and prints “Hello”. Easy, right? Each of these steps is needed to make sure communication can still happen even if a peer disconnects or reconnects at a different network address.
+                </p>
```
Comment (jefft0) [inline pages/posts/flight-of-a-byte.tsx:82]: I'm not sure why client2 is in code font but client1 isn't. Would it be possible to be consistent?

**Review by jefft0 — CHANGES_REQUESTED** [review]

File: pages/posts/flight-of-a-byte.tsx (hunk 10 above)
Comment (iuricmp) [reply pages/posts/flight-of-a-byte.tsx:81]: Sorry, I forgot to include the links. I just added them and I think we are good now.

File: pages/posts/flight-of-a-byte.tsx (hunk 11 above)
Comment (iuricmp) [reply pages/posts/flight-of-a-byte.tsx:85]: done.

File: pages/posts/flight-of-a-byte.tsx [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+                <h3>libp2p pubsub</h3>
+                <p>
+                 So now the question is, how does <code>client1</code> learn that <code>client2</code> says there is a new message log head? This uses libp2p pubsub where peers can subscribe to a topic and publish a pubsub message to that topic, or receive pubsub messages. (We say “pubsub message” to not confuse with the ordinary messages in the log which we&apos;ll discuss below.)
+                </p>
+                <p>
+                libp2p pubsub is a powerful tool with algorithms which allow it to scale to many peers and to make sure they all get the pubsub message. You can read more in <a href="https://docs.libp2p.io/concepts/pubsub/overview/" target="_blank" rel="noreferrer">its documentation</a>, but for this blog post we only need to say that the Wesh group has a libp2p topic name like 
+                &nbsp;<code>
+                /orbitdb/bafyreihpdptqwgpuwu4ob6nusxebmdx6fm6mq5pma5hva6eu5u74pflo7e/2ac4e88c5272f900c76fd36989f9780e5b2c95d75d38fe5bcef0345b34bc4806_wesh_group_messages
+                    </code>
```
Comment (LFGaming) [inline pages/posts/flight-of-a-byte.tsx:65]: There is a space before /orbitdb/...
```suggestion
                &nbsp;
                <code> /orbitdb/bafyreihpdptqwgpuwu4ob6nusxebmdx6fm6mq5pma5hva6eu5u74pflo7e/2ac4e88c5272f900c76fd36989f9780e5b2c95d75d38fe5bcef0345b34bc4806_wesh_group_messages
                    </code>
```

File: pages/posts/flight-of-a-byte.tsx [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+                </p>
+                <h3>Flight of a Byte</h3>
+                <p>
+                In summary, the flight of a message from <code>client2</code> to <code>client1</code> starts when the <code>client2</code> peer is discovered, subscribes to the libp2p pubsub topic for the group messages and publishes its message log head. <code>client1</code> doesn&apos;t have this head yet, so it uses the CID to fetch the message using IPFS. When the message arrives, it is added to <code>client1&apos;s</code> copy of the message log in the correct order. Wesh sends this new message to processes using the <code>GroupMessageList</code> service, such as the sample application which receives it and prints “Hello”. Easy, right? Each of these steps is needed to make sure communication can still happen even if a peer disconnects or reconnects at a different network address.
```
Comment (LFGaming) [inline pages/posts/flight-of-a-byte.tsx:84]: CID was in code blocks everywhere but here.


```suggestion
                In summary, the flight of a message from <code>client2</code> to <code>client1</code> starts when the <code>client2</code> peer is discovered, subscribes to the libp2p pubsub topic for the group messages and publishes its message log head. <code>client1</code> doesn&apos;t have this head yet, so it uses the <code>CID</code> to fetch the message using IPFS. When the message arrives, it is added to <code>client1&apos;s</code> copy of the message log in the correct order. Wesh sends this new message to processes using the <code>GroupMessageList</code> service, such as the sample application which receives it and prints “Hello”. Easy, right? Each of these steps is needed to make sure communication can still happen even if a peer disconnects or reconnects at a different network address.
```

File: pages/posts/flight-of-a-byte.tsx [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+                &nbsp;<code>
+                /orbitdb/bafyreihpdptqwgpuwu4ob6nusxebmdx6fm6mq5pma5hva6eu5u74pflo7e/2ac4e88c5272f900c76fd36989f9780e5b2c95d75d38fe5bcef0345b34bc4806_wesh_group_messages
+                    </code>
+                &nbsp;a pubsub message to this topic that contains the message log head. The libp2p on <code>client1</code> receives this pubsub message and sends it (<a href="https://github.com/libp2p/go-libp2p-pubsub/blob/6d73cd4b56bf9d7fdd0c0caaed629a3d51027a75/pubsub.go#L978" target="_blank" rel="noreferrer">code</a>) internally to all processes which are subscribed to the topic. The Wesh application of <code>client1</code> has used ActivateGroup and is therefore subscribed to this topic and receives the pubsub message (<a href="https://github.com/berty/go-orbit-db/blob/368618e06b3be3bd17952d167d0e2f4fa03960ff/pubsub/pubsubraw/pubsub.go#L76" target="_blank" rel="noreferrer">code</a>).
```
Comment (LFGaming) [inline pages/posts/flight-of-a-byte.tsx:66]: GroupMessageList is in a code block, but ActivateGroup not.

File: pages/posts/flight-of-a-byte.tsx (hunk 6 above)
Comment (jefft0) [reply pages/posts/flight-of-a-byte.tsx:67]: Still need to fix this:
"a pubsub message to this topic" should be "Client2 publishes a pubsub message to this topic."

File: pages/posts/flight-of-a-byte.tsx (hunk 14 above)
Comment (jefft0) [reply pages/posts/flight-of-a-byte.tsx:66]: I agree. For consistency, can you put ActivateGroup in code font.

**Comment by jefft0** [comment]
Almost finished. We only need the fix mentioned above:
"a pubsub message to this topic" should be "Client2 publishes a pubsub message to this topic."

**Review by jefft0 — APPROVED** [review]

**Pull request merged**